### PR TITLE
Fix percent decoding and normalization

### PR DESCRIPTION
**Pull request by mahmoud**
Per #16, hyperlink's inherited behavior was applying percent-decoding too broadly. Each percent-encodable field in the URL (userinfo, path, query string, and fragment), has certain special characters that should *never* be decoded [1].

To this end, this PR adds and utilizes a `_decode_XXX_part()` function for each of the fields, symmetrical to the `_encode_XXX_part()` that was already there.

Several test cases were added, as well as a general test for the invariant that multiple calls of `to_iri()` will not generate different URL objects.

@glyph has shown the most interest in this feature, so I've added him as reviewer.

[1]: To take the path as an example, the special characters include delimiters, such as the slash (`/`), question mark (`?`), and hash mark (`#`), as well as the percent sign (`%`) itself.

### CODE REVIEW

File: hyperlink/test/test_url.py [hunk 1]
```diff
@@ -115,6 +115,29 @@
...
+    'https://example.com/%3F/',  # question mark in path
+    'https://example.com/%23/',  # hash in path
+    'https://example.com/?%23=b',  # hash in query param name
+    'https://example.com/?%3D=b',  # equals in query param name
+    'https://example.com/?%26=b',  # ampersand in query param name
+    'https://example.com/?a=%23',  # hash in query param value
+    'https://example.com/?a=%26',  # ampersand in query param value
+    'https://example.com/?a=%3D',  # equals in query param value
+    # double-encoded percent sign in all percent-encodable positions:
+    "http://(%2525):(%2525)@example.com/(%2525)/?(%2525)=(%2525)#(%2525)",
```
Comment (glyph) [inline hyperlink/test/test_url.py]: 👍

File: hyperlink/test/test_url.py [hunk 2]
```diff
@@ -231,9 +254,21 @@ def test_roundtrip(self):
         L{URL.to_text} should invert L{URL.from_text}.
         """
         for test in ROUNDTRIP_TESTS:
-            result = URL.from_text(test).to_text()
+            result = URL.from_text(test).to_text(with_password=True)
             self.assertEqual(test, result)
 
+    def test_roundtrip_double_iri(self):
```
Comment (glyph) [inline hyperlink/test/test_url.py]: This looks like a place where it might be interesting to drop in hypothesis (since it was in fact hypothesis that caught this bug, in txacme's test suite)

File: hyperlink/_url.py [hunk 3]
```diff
@@ -861,6 +917,8 @@ def replace(self, scheme=_UNSET, host=_UNSET, path=_UNSET, query=_UNSET,
             port=_optional(port, self.port),
             rooted=_optional(rooted, self.rooted),
             userinfo=_optional(userinfo, self.userinfo),
+            family=_optional(family, self.family),
+            uses_netloc=_optional(uses_netloc, self.uses_netloc)
```
Comment (glyph) [inline hyperlink/_url.py]: This seems incorrect.  You shouldn't be able to specify `uses_netloc` to the constructor at all, it's entirely a function of the scheme; adding it to `replace` seems to be propagating the error.

Comment (glyph) [inline hyperlink/_url.py]: Confirming this, deleting these added lines doesn't make any tests fail; if it should be here at all, it shouldn't be in this change.

File: hyperlink/test/test_url.py (hunk 2 above)
Comment (mahmoud) [inline hyperlink/test/test_url.py]: yeah, i'm down for some hypothesis in the future. i played with it a bit on this PR, but haven't delved deep enough yet.

File: hyperlink/_url.py (hunk 3 above)
Comment (mahmoud) [inline hyperlink/_url.py]: Functionally, `uses_netloc` is there because for unrecognized schemes parsed using `from_text`, URL will persist whether or not the original URL used the netloc slashes.

Logistically, I can't remember if I intended to include it in this PR or if I just saw the big regression of not passing through `family` and `uses_netloc` -- both in the constructor and the docstring of `replace`, but missing in the function signature itself -- and fixed it post haste.

Comment (glyph) [inline hyperlink/_url.py]: If it's worth working it's worth testing :-). Urgency notwithstanding it seems like an unrelated change.

Comment (mahmoud) [inline hyperlink/_url.py]: Well, the branch started out as seeking to normalize decoding in general, but it became pretty percent-decoding centric, as idna improvements will come in another PR. 

I've been brainstorming other ways to go with some of these latter-day arguments, but for now I think I prefer bringing things into sync, so I've added a test.

**Comment by glyph** [comment]
OK, I guess the documentation already indicates that this is possible.

That said: has a release gone out with this parameter in it?  I think it's an ugly wart on the interface and I'd prefer to find a different way to deal with this if we can before committing to public API for it.

File: hyperlink/_url.py [hunk 4]
```diff
@@ -160,6 +152,19 @@ def __nonzero__(self):
 _QUERY_DELIMS = _ALL_DELIMS - _QUERY_SAFE
 
 
+def _make_decode_map(delims, allow_percent=False):
+    ret = dict(_HEX_CHAR_MAP)
+    if not allow_percent:
+        delims = set(delims) | set([u'%'])
+    for delim in delims:
+        _hexord = hex(ord(delim))[2:].zfill(2).encode('ascii').upper()
```
Comment (alexwlchan) [inline hyperlink/_url.py]: minor: I think you could simplify this by using the `%x` format tokens:

```python
_hexord = format(ord(delim), '02X').encode('ascii')
```

(I don’t know if there’s a performance difference, or if this code is particularly perf-critical)

Comment (mahmoud) [inline hyperlink/_url.py]: This code is executed 4x at startup. Each time the loop happens <10 times. ;)

Comment (glyph) [inline hyperlink/_url.py]: It's shorter and easier to read, so the performance issue may be secondary :)

Comment (glyph) [inline hyperlink/_url.py]: In any case a follow-up PR to do this would be fine.

**Comment by mahmoud** [comment]
@glyph yes, it's had this parameter from the initial release. It's deep enough down the parameter list that practically I don't consider it a public API. That said, with an immutable object like this, there's no other way to control whether the slashes should be emitted.

If you're wanting to slim down the parameter list, I have an idea on how to get rid of `family`... For another PR of course :)
